### PR TITLE
Directly execute ginkgo in a test job

### PR DESCRIPTION
**Pull request by justinsb**
Create a test job that directly executes ginkgo (triggered by a
temporary magic env-var), and teach kubetest to call ginkgo directly.

Issue #3330

### CODE REVIEW

**Comment by justinsb** [comment]
cc @chrislovecnm 

Not yet tested

File: kubetest/e2e/runner.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	NodeOSDistribution    string
+	NumNodes              int
+	ReportDir             string
+	ReportPrefix          string
+}
+
+func NewTester() *Tester {
+	t := &Tester{
+		FlakeAttempts: 1,
+		GinkgoPath:    "ginkgo",
```
Comment (chrislovecnm) [inline kubetest/e2e/runner.go]: This changes if we are in bazel build or go build. Can we just do a bazel run???

**Comment by chrislovecnm** [comment]
You also might need to node instance group name(s) in gce

**Comment by justinsb** [comment]
In my local testing this now gets the kops-on-GCE tests running again (they are currently broken https://k8s-testgrid.appspot.com/google-kops-gce ).  There are still ~11 individual test failures, some of which are due to the ginkgo setup, some of which are because we weren't previously passing provider=gce.  My preference is to work through those in separate PRs.

What do we think about executing ginkgo directly (replacing the `ginkgo-e2e.sh` script)?  It should move us forwards on #3330 cc @krzyzacy @fejta @BenTheElder

**Comment by krzyzacy** [comment]
yup that's a good direction

File: jobs/config.json [hunk 2]
```diff
@@ -9884,6 +9884,25 @@
       "sig-aws"
     ]
   },
+  "ci-kubernetes-e2e-kops-gce-ginkgodirect": {
+    "args": [
+      "--cluster=e2e-kops-gce-ginkgodirect.k8s.local",
+      "--deployment=kops",
+      "--env=KUBETEST_RUN_GINKGO_DIRECT=y",
```
Comment (krzyzacy) [inline jobs/config.json]: make that a flag?

Comment (justinsb) [reply jobs/config.json]: Happy to do that if you think so.  Here was my thought process though:

* Setting up the flags for ginkgo needs provider-specific logic (generally)  i.e. kops & kube-anywhere would have different flags.  The ginkgo-e2e.sh logic holds for kube-up, but not for kops, for example.
* I don't see a provider supporting multiple runners (both ginkgo direct and ginkgo-e2e.sh) long-term
* The env var I see as a temporary thing, until kops can get 100% to using ginkgo-direct.  I don't want to break the kops-on-AWS tests while I get it working though.
* Once it is working, I would just remove the env var.

I guess we could also just drop the flag in future though, but that's more of a breaking change...

**Comment by BenTheElder** [comment]
cc @Q-Lee

File: jobs/config.json (hunk 2 above)
Comment (BenTheElder) [reply jobs/config.json]: a flag is a lot less magical to understand :-)
for the --env flag here I need to know what reads this, how it converts to env, and what reads the env.
for a flag I just see the flag here and the corresponding definition in kubetest.

File: kubetest/e2e.go [hunk 3]
```diff
@@ -230,9 +231,17 @@ func run(deploy deployer, o options) error {
...
-						errs = util.AppendError(errs, control.XmlWrap(&suite, "Test", func() error {
-							return test(testArgs)
-						}))
+						var tester Tester
+						tester = &GinkgoScriptTester{}
+						if testBuilder, ok := deploy.(testBuilder); ok {
+							tester, err = testBuilder.BuildTester(&o)
+							errs = util.AppendError(errs, err)
+						}
+						if tester != nil {
```
Comment (krzyzacy) [inline kubetest/e2e.go]: we'd probably always fallback to `GinkgoScriptTester`

File: kubetest/e2e/runner.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+	"k8s.io/test-infra/kubetest/process"
+)
+
+// Tester runs e2e tests directly (by calling ginkgo)
+type Tester struct {
+	// Required
+	Kubeconfig string
+	Provider   string
+	KubeRoot   string
```
Comment (krzyzacy) [inline kubetest/e2e/runner.go]: validate those required fields are not empty

File: kubetest/kops.go [hunk 5]
```diff
@@ -530,6 +531,40 @@ func (k kops) TestSetup() error {
 	return nil
 }
 
+// BuildTester returns a standard ginkgo-script tester, unless KUBETEST_RUN_GINKGO_DIRECT=y, when we build an e2e.Tester
+func (k kops) BuildTester(o *options) (Tester, error) {
```
Comment (krzyzacy) [inline kubetest/kops.go]: Why this should live in kops.go? Looks like it can be shared across multiple deployers..

**Comment by Q-Lee** [comment]
The kubetest/dind/dind.go does something similar. We should consider refactoring common functionality?

File: jobs/config.json (hunk 2 above)
Comment (Q-Lee) [reply jobs/config.json]: I agree. We're trying to unify and clean-up the flags in kubetest.

File: kubetest/kops.go (hunk 5 above)
Comment (justinsb) [reply kubetest/kops.go]: This is building the settings that kops sets for calling ginkgo.  My opinion is that we should not share this, even if we end up with similar implementations in each.  The reason is that we have a shared implementation today, in bash, and it is hard to make the changes needed in kops, and IIRC the bash has lots of conditionals to enable it to be used elsewhere.

I'd suggest it's probably simpler to just go with simpler explicit logic, even at the cost of duplicating.  If we then decide to pull out a common "initialize" function, we could do that - but I suspect it won't be worth it.

File: kubetest/e2e.go (hunk 3 above)
Comment (justinsb) [reply kubetest/e2e.go]: So we do, unless BuildTester returns nil - we then assume that running the ginkgo tests is not appropriate.  `BuildTester` could always itself return the ginkgo tester in the case of error, but I think the right thing to do here is to fail the test.

Another way to say this: the fallback is for configurations that haven't implemented BuildTester, not for configurations where trying to build a tester object encounters an error.

**Comment by justinsb** [comment]
Removed the magic env var - we just run it on kops/GCE instead.  I really don't want to create a flag that is so short-lived, and kops on GCE is broken anyway.

Also refactored conformance and dind to use the ginkgo runner.  PTAL @Q-Lee

**Comment by justinsb** [comment]
What do we think?  Can we get this in?  Should I revert it so that it just affects kops-on-GCE (for now)?

File: kubetest/e2e.go [hunk 6]
```diff
@@ -732,6 +717,20 @@ func skewTestEnv(env, args []string, prefix string, checkSkew bool) error {
 	return control.FinishRunning(cmd)
 }
 
-func test(testArgs []string) error {
+// GinkgoScriptTester implements Tester by calling the hack/ginkgo-e2e.sh script
+type GinkgoScriptTester struct {
```
Comment (krzyzacy) [inline kubetest/e2e.go]: should this be in the e2e package?

**Comment by krzyzacy** [comment]
lgtm overall, punt to @BenTheElder for dind changes

File: kubetest/e2e.go (hunk 6 above)
Comment (justinsb) [reply kubetest/e2e.go]: I was trying to keep the diff simpler.  Happy to move it if you prefer, but my hope is that this goes away...

**Comment by justinsb** [comment]
Thanks @krzyzacy!

@BenTheElder what do you think?  I can revert the dind changes for now (or split them out) if you'd prefer.

**Comment by BenTheElder** [comment]
taking a look, I'm preparing to rework the dind stuff majorly so i'm just going to make sure it looks good for kubetest and not worry too much about dind.

File: kubetest/conformance/conformance.go [hunk 7]
```diff
@@ -42,78 +40,44 @@ type Tester struct {
...
-	skipArg := fmt.Sprintf("--skip=%s", skipRegex)
-
-	// Execute ginkgo, which in turn executes e2e.test.
-	args := []string{"--seed=1436380640", "--nodes=10", focusArg, skipArg, t.e2etest,
-		"--", "--kubeconfig", t.kubecfg, "--ginkgo.flakeAttempts=2", "--num-nodes=4", "--systemd-services=docker,kubelet",
-		"--report-dir", t.reportdir}
-	args = append(args, strings.Fields(*t.testArgs)...)
-	cmd := exec.Command(t.ginkgo, args...)
-	return t.control.FinishRunning(cmd)
+	t.Seed = 1436380640
```
Comment (BenTheElder) [inline kubetest/conformance/conformance.go]: is this normally fixed in the existing runners?

File: kubetest/e2e.go [hunk 8]
```diff
@@ -190,48 +190,33 @@ func run(deploy deployer, o options) error {
...
 			}))
-		} else if o.deployment == "dind" {
-			if err := control.XmlWrap(&suite, "IsUp", deploy.IsUp); err != nil {
-				errs = util.AppendError(errs, err)
-			}
-			dep := deploy.(*dind.DindDeployer)
-			tester, err := dep.NewTester(o.focusRegex, o.skipRegex, o.ginkgoParallel.Get())
-			if err != nil {
-				return err
+		} else {
```
Comment (krzyzacy) [inline kubetest/e2e.go]: @justinsb awwwwwww else if here!

**Pull request merged**